src/analysis/processing/qgsalgorithmgrid.cpp

```diff
@@ -0,0 +1,527 @@
+
+  if( mGridExtent.width() < mHSpacing ) //check if grid extent is smaller than horizontal spacing
+    throw QgsProcessingException( QObject::tr( "Horizontal spacing is too large for the covered area."));
+
+  if( mGridExtent.height() < mVSpacing) //check if grid extent is smaller than vertical spacing
+    throw QgsProcessingException( QObject::tr( "Vertical spacing is too large for the covered area."));
+
+  if( mHSpacing <= mHOverlay or mVSpacing <= mVOverlay)
+    throw QgsProcessingException( QObject::tr( "Invalid overlay: horizontal: '%1', vertical: '%2'").arg( mHOverlay ).arg( mVOverlay ));
+
+  //add grid fields
+  QgsFields fields = QgsFields();
+  fields.append(QgsField(QStringLiteral( "id "), QVariant::Int));
+  fields.append(QgsField(QStringLiteral( "left" ), QVariant::Double));
+  fields.append(QgsField(QStringLiteral( "top" ), QVariant::Double));
+  fields.append(QgsField(QStringLiteral( "right" ), QVariant::Double));
+  fields.append(QgsField(QStringLiteral( "bottom" ), QVariant::Double));
+
+  //define output WKB-type
+  QgsWkbTypes::Type outputWkb;
+  if( mIdx == 0 )
+  {
+    outputWkb = QgsWkbTypes::Point;
+  }
+  else if ( mIdx == 1 )
+  {
+    outputWkb = QgsWkbTypes::LineString;
+  }
+  else
+  {
+    outputWkb = QgsWkbTypes::Polygon;
+  }
+
+  //output sink
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, outputWkb, mCrs ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  feedback->setProgress(0);
+  //select calculation method based on grid type selection
+  switch ( mIdx )
+  {
+    case 0: //point
+      createPointGrid(sink, feedback);
+      break;
+    case 1: //line
+      createLineGrid(sink, feedback);
+      break;
+    case 2: //rectangle
+      createRectangleGrid(sink, feedback);
+      break;
+    case 3: //diamond
+      createDiamondGrid(sink, feedback);
+      break;
+    case 4: //hexagon
+      createHexagonGrid(sink, feedback);
+      break;
+    default:
+      break; //no default needed
+  }
+
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+  return outputs;
+}
+
+void QgsGridAlgorithm::createPointGrid(std::unique_ptr< QgsFeatureSink > &sink, QgsProcessingFeedback *feedback)
+{
+  QgsFeature f = QgsFeature();
+
+  int cols =  static_cast<int>( std::ceil(mGridExtent.width()) / (mHSpacing - mHOverlay) );
+  int rows = static_cast<int>( std::ceil(mGridExtent.height()) / (mVSpacing - mVOverlay) );
+
+  int id = 1;
+  int cnt = 0;
+  int cellcnt = rows * cols;
+  double cnt_log = static_cast<double>(cellcnt) * 0.01;
+
+  for(int col = 0; col < cols; col++)
+  {
+    for(int row = 0; row < rows; row++)
+    {
+      double x = mGridExtent.xMinimum() + (col * mHSpacing - col * mHOverlay);
```

Comment:
This could potentially be moved up into the outer loop, as we don't need to calculate it for every row.